feat(api): Add healthcheck field to Config

bacongobbler opened the pull request:
# Summary of Changes

This PR adds a `healthcheck` field to the controller to handle the `deis healthchecks` command, and to eventually deprecate the old way of setting healthchecks via `deis config:set HEALTHCHECK_URL`. This is a continuation of #823.

# Issue(s) that this PR Closes

closes #603
closes #599

# Associated [End To End Test](https://github.com/deis/workflow-e2e) PR(s)

TODO

# Associated [Documentation](https://github.com/deis/workflow) PR(s)

deis/workflow#322

# Associated [Design Document](http://docs.deis.io/en/latest/contributing/design-documents)(s)

deis/workflow#322

# Testing Instructions

TODO

# Pull Request Hygiene TODOs

Please make sure the below checklist is complete.

- [x] Your pull request is concise and to the point (make another PR for refactoring nearby code)
- [x] Your commits are squashed into logical units of work
- [x] Your commits follow the [commit style guidelines](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide)

:cherry_blossom: Thank you! :cherry_blossom:

bacongobbler commented:
so for others to be aware of, this PR passes when setting `exec` and `tcpSocket` checks, but for some reason I cannot set `httpGet` probes... Will have to look more into it, because all I get back from the server is

```
><> deis healthchecks:set readiness httpGet 5000
Applying readinessProbe healthcheck... Error: 400 Bad Request

><> kd logs deis-controller-abc12
10.246.72.17 "POST /v2/apps/go/config/ HTTP/1.1" 400 47 "Deis Client v2.1.0-dev"
```

No traceback :cry:

helgi commented:
I believe you are being hit by https://github.com/deis/controller/issues/816

helgi commented:
Can we change the migration so that `HEALTHCHECK_` gets migrated from `values` into `healthchecks` with it's appropriate defaults under `httpGet` ?

Same basically needs to be done when people do `config:set HEALTHCHECK*`

Doing that would reduce the various duplications in the code that spider outside of the `Config` model

bacongobbler commented:
> Can we change the migration so that HEALTHCHECK_ gets migrated from values into healthchecks with it's appropriate defaults under httpGet ?

This might be easier to tackle via the client. We can convert HEALTHCHECK_* in the client to send a real `healthcheck` object, but keep this code in for old time's sake until we eventually deprecate it. I'm not sure we can easily do a data migration from config to healthcheck, and we still need to support existing `config` objects that handle HEALTHCHECK_*. would that be reasonable, or how would you imagine we could tackle this issue?

helgi commented:
We can't really guarantee that a user is on the latest client, right?

bacongobbler commented:
also what if for some crazy reason the user also required that envvar in their app? Migrating it to the healthcheck field would mean the envvar is no longer present in their app (considered a breaking change)

helgi commented:
I think that might be a stretch that in the 2.0 timeframe someone would start depending on the healthcheck env var within their own app

helgi commented:
Another option is to populate `healthchecks` but also keep it in `values` that way the env var stays but internally we always use `healthchecks`

bacongobbler commented:
that would be a good tradeoff. I can tackle that, most likely.

bacongobbler commented:
this is finally ready for review. @helgi I implemented the change you requested and added unit tests to cover that. Now any time a config object is saved (including existing ones), a healthcheck object is created if the legacy envvars exist.

helgi commented on rootfs/scheduler/__init__.py:
```diff
@@ -936,62 +936,6 @@ def delete_rc(self, namespace, name):
 
         return response
 
-    def _healthcheck(self, namespace, container, routable=False, path='/', port=5000,
```
I find it a little dangerous when we are not keeping the healthcheck logic inside the scheduler but it is rather now outside in serializer class... Thoughts?

bacongobbler replied:
this is essentially doing the same job as `jsonschema` is in the serializer, but it fails much later in the process *after* someone has already created the healthcheck. I think the way we validate the healthcheck through the serializer is much better, but I'm open to opinions.

helgi replied:
Yes, it sure is and I think the serializer should still do the validation but I was mostly wondering given how we consider the scheduler a separate module that maybe serializer should pull in "validation" from the scheduler plus we may have a different API vs Scheduler structure in the health check department, either due to differences in k8s versions or changes in internal design.

Trying to make sure we don't cause us debt later :) I'm not fully sold on any of the above ideas particularly, yet.

bacongobbler replied:
so you want to move the k8s healthcheck schema over to the `scheduler` module instead of sticking it in the serializer? That makes sense. Just trying to grok what you wanna change :)

helgi replied:
Yeah, that was **one** of the things rattling around. Was also looking for other ideas if you see a better way :) A bit of pre-caffeine rubber ducking.

If we moved it over then I wouldn't want to lose the error message handling in the Serializer. It's a little funky when we have such big scope shared in two places but the nature of the beast I guess :)

helgi commented on rootfs/api/models/config.py:
```diff
@@ -137,13 +111,12 @@ def save(self, **kwargs):
                 setattr(self, attr, data)
 
-            self.set_healthchecks()
+            self._migrate_legacy_healthcheck()
```
Nice! We just need to aggressively ensure that people don't have legacy health check and then try to set new checks. On controller restart it would always overwrite. Maybe add in a "if livenessProbe in self.healthcheck" is true then skip setting legacy

bacongobbler replied:
Good thinking. My current logic is that if someone is using legacy healthchecks, they're gonna have to unset them before continuing on to set a livenessProbe healthcheck.

Shall I also add a deprecation notice here to eventually rip this out after 2.4 or something?

helgi replied:
Yeah, I think we may have to keep it to the next major release given how strict we are usually on breakages. @slack thoughts on how long we should keep it deprecated before removal?

kmala replied:
The reason I din't migrate before I left is how are we gonna handle port.... current environment health checks doesn't provide port and we use to get it from releases in the scheduler logic.... Port is the only mandatory field and we won't have handle of it in the config object

helgi replied:
We could pull the service port for now to solve for that

bacongobbler replied:
The existing behaviour of pulling in the service port should work in that case. I'll see if I can write some tests for that.

kmala replied:
You can set config before creating a build and the service port is actually overridden during deploy of the app but the migration happens during a new config....
What-if the client does all his config first and does a build next...then the health check might have wrong port

bacongobbler replied:
I fixed this in 6fd0add by only injecting the port if it is not already defined (such as when using `deis config:set HEALTHCHECK_URL`). That should alleviate your concern with 

> What-if the client does all his config first and does a build next...then the health check might have wrong port

bacongobbler commented on rootfs/api/tests/test_config.py:
```diff
@@ -849,3 +849,125 @@ def test_healthchecks_validations(self, mock_requests):
+        self.assertEqual(response.status_code, 201, response.data)
+        self.assertIn('HEALTHCHECK_URL', response.data['values'])
+        self.assertEqual(response.data['values']['HEALTHCHECK_URL'], '/health')
+        # legacy defaults
```
A note I should bring up... Are these really the defaults we should be using? I referenced them based off of https://github.com/deis/controller/blob/master/rootfs/api/models/config.py#L40-L45 but the scheduler has conflicting options.

helgi replied:
Yeah the Model always overwrote the defaults in the scheduler, the former may have evolved while the latter was just left behind with an egg on its face

bacongobbler commented:
I am not proud of this code, but https://github.com/deis/controller/pull/836/commits/6fd0add9908272116500beb540676386bdd968cd should do it.

helgi commented:
I added a `liveness` probe of `httpGet` to an app (which worked) and then later I added a `readiness` probe of `exec` and I got this exceptions:

```
ERROR [oddest-lungfish]: oddest-lungfish-v18-cmd (app::deploy): 'livenessProbe'
ERROR:root:oddest-lungfish-v18-cmd (app::deploy): 'livenessProbe'
Traceback (most recent call last):
  File "/app/scheduler/__init__.py", line 114, in deploy
    self.get_rc(namespace, name)
  File "/app/scheduler/__init__.py", line 709, in get_rc
    'get ReplicationController "{}" in Namespace "{}"', name, namespace
scheduler.KubeHTTPException: ('failed to get ReplicationController "oddest-lungfish-v18-cmd" in Namespace "oddest-lungfish": 404 Not Found\n{\'message\': \'replicationcontrollers "oddest-lungfish-v18-cmd" not found\', \'kind\': \'Status\', \'reason\': \'NotFound\', \'code\': 404, \'status\': \'Failure\', \'apiVersion\': \'v1\', \'metadata\': {}, \'details\': {\'name\': \'oddest-lungfish-v18-cmd\', \'kind\': \'replicationcontrollers\'}}', 'oddest-lungfish-v18-cmd', 'oddest-lungfish')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/api/models/app.py", line 445, in deploy
    **kwargs
  File "/app/scheduler/__init__.py", line 120, in deploy
    new_rc = self.create_rc(namespace, name, image, command, replicas=0, **kwargs).json()
  File "/app/scheduler/__init__.py", line 889, in create_rc
    manifest['spec']['template'] = self._build_pod_manifest(namespace, name, image, **kwargs)
  File "/app/scheduler/__init__.py", line 348, in _build_pod_manifest
    self._set_container(namespace, container_name, container, **kwargs)
  File "/app/scheduler/__init__.py", line 516, in _set_container
    healthchecks['livenessProbe'].get('httpGet') is not None and
KeyError: 'livenessProbe'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/api/views.py", line 261, in post_save
    config.app.deploy(self.release)
  File "/app/api/models/app.py", line 457, in deploy
    raise ServiceUnavailable(err) from e
api.exceptions.ServiceUnavailable: oddest-lungfish-v18-cmd (app::deploy): 'livenessProbe'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/rest_framework/views.py", line 463, in dispatch
    response = handler(request, *args, **kwargs)
  File "/app/api/views.py", line 171, in create
    return super(AppResourceViewSet, self).create(request, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/rest_framework/mixins.py", line 21, in create
    self.perform_create(serializer)
  File "/app/api/viewsets.py", line 21, in perform_create
    self.post_save(obj)
  File "/app/api/views.py", line 264, in post_save
    raise DeisException(str(e)) from e
api.exceptions.DeisException: oddest-lungfish-v18-cmd (app::deploy): 'livenessProbe'
```

bacongobbler commented:
this was actually part of the fix I made this morning for the auto-port populate. thanks for catching that!

bacongobbler commented:
fix'd
